subcommand/acl-controller/command_test.go

```diff
@@ -3,22 +3,307 @@ package aclcontroller
+				cmd.flagPartition = testPartitionName
+			}
+
+			// Upsert once to create the resources
+			err = cmd.upsertConsulResources(consulClient, testTaskMetadataResponse)
+			require.NoError(t, err)
+
+			checkConsulResources(t, consulClient, c.expPolicyRules, c.partitionsEnabled)
+
+			// Optionally, delete some of the resources
+			if c.deleteAuthMethods {
+				methods, _, err := consulClient.ACL().AuthMethodList(nil)
+				require.NoError(t, err)
+				for _, method := range methods {
+					_, err := consulClient.ACL().AuthMethodDelete(method.Name, nil)
+					require.NoError(t, err)
+				}
+				methods, _, err = consulClient.ACL().AuthMethodList(nil)
+				require.NoError(t, err)
+				require.Len(t, methods, 0)
+			}
+			if c.deletePolicy {
+				policies, _, err := consulClient.ACL().PolicyList(nil)
+				require.NoError(t, err)
+				for _, policy := range policies {
+					if policy.Name != "global-management" {
+						_, err := consulClient.ACL().PolicyDelete(policy.ID, nil)
+						require.NoError(t, err)
+					}
+				}
+				// Only the global management policy should exist.
+				policies, _, err = consulClient.ACL().PolicyList(nil)
+				require.NoError(t, err)
+				if c.partitionsEnabled {
+					require.Len(t, policies, 0)
+				} else {
+					// The default partition has a global-management policy
+					require.Len(t, policies, 1)
+					require.Equal(t, policies[0].Name, "global-management")
+				}
+			}
+			if c.deleteRole {
+				roles, _, err := consulClient.ACL().RoleList(nil)
+				require.NoError(t, err)
+				for _, role := range roles {
+					_, err := consulClient.ACL().RoleDelete(role.ID, nil)
+					require.NoError(t, err)
+				}
+				roles, _, err = consulClient.ACL().RoleList(nil)
+				require.NoError(t, err)
+				require.Len(t, roles, 0)
+			}
+			if c.deleteBindingRules {
+				methods, _, err := consulClient.ACL().AuthMethodList(nil)
+				require.NoError(t, err)
+				for _, method := range methods {
+					rules, _, err := consulClient.ACL().BindingRuleList(method.Name, nil)
+					require.NoError(t, err)
+					for _, rule := range rules {
+						_, err := consulClient.ACL().BindingRuleDelete(rule.ID, nil)
+						require.NoError(t, err)
+					}
+					rules, _, err = consulClient.ACL().BindingRuleList(method.Name, nil)
+					require.NoError(t, err)
+					require.Len(t, rules, 0)
+				}
+			}
+			if c.partitionsEnabled && c.deletePartition {
+				ctx := context.Background()
+				partitions, _, err := consulClient.Partitions().List(ctx, nil)
+				require.NoError(t, err)
+				for _, part := range partitions {
+					if part.Name != "default" {
+						_, err := consulClient.Partitions().Delete(ctx, part.Name, nil)
+						require.NoError(t, err)
+					}
+				}
+				// Apparently, takes moment for the partition to actually be deleted.
+				retry.RunWith(&retry.Timer{Timeout: 2 * time.Second, Wait: 200 * time.Millisecond}, t, func(r *retry.R) {
+					partitions, _, err = consulClient.Partitions().List(ctx, nil)
```

Comment:
I believe the first sign that a partition is being deleted is `DeletedAt` being set on partitions.

Reply:
Yeah. Would that change the need for a retry loop like this though? I don't think so.

Reply:
Not that it's worth doing, but I think we could eliminate the retry loop by ensuring that the length of `partitions` is one after removing partitions with `DeletedAt` set. That makes the code less understandable so I don't think it's worth it.

Reply:
Yeah, I think in that case we'd also need to change the acl-controller code to look for `DeletedAt` as well, and even then I'm not sure if, while the partition is "present" but has `DeletedAt` set, that we could create a partition with the same name?

I think I agree - probably not worth it, but super helpful to think through!